### PR TITLE
Added `ConjArray` wrapper type for conjugate array views

### PR DESCRIPTION
**Pull request by andyferris**
By default, this is used only in conjugation with `RowVector`, so that both `transpose(vec)` and `ctranspose(vec)` both return views.

However, this will pave the way for views for transposed matrices and replacement of `Ac_mul_B`, etc by specialized methods on `*`, simplifying parsing.

### CODE REVIEW

**Comment by andyferris** [comment]
Follow-up of #19670

File: base/linalg/conjarray.jl
Comment (stevengj) [inline base/linalg/conjarray.jl]: Seems like it should pass through a `strides` and `pointer` (and `unsafe_convert` to a pointer) call to the underlying array.  (Similarly for `RowVector`.)

File: base/linalg/conjarray.jl
Comment (stevengj) [inline base/linalg/conjarray.jl]: We should probably have a `similar` function that calls `similar(parent)`.

File: base/linalg/conjarray.jl
Comment (stevengj) [inline base/linalg/conjarray.jl]: Don't think this method is needed.

**Comment by stevengj** [comment]
Should't there be some more specialized linear-algebra operations?  e.g. `x' * y` should turn into a `dot` call.   `_conj(a) + _conj(b)` should be the same as `_conj(a + b)` (similarly for `*`).

File: NEWS.md
Comment (stevengj) [inline NEWS.md]: `copies.` -> `copies ([#20047]).`

File: base/linalg/conjarray.jl
Comment (stevengj) [inline base/linalg/conjarray.jl]: Actually, not `pointer` since the underlying data is not conjugated.  But we should provide `pointer` for `RowVector`.  And we should still provide `stride` for `ConjArray` (this is potentially useful for telling you what direction to traverse an array).

File: test/sparse/sparse.jl
Comment (tkelman) [inline test/sparse/sparse.jl]: this should be replaced with a test that serves the same purpose, if possible, cc @sacha0

Comment (andyferris) [inline test/sparse/sparse.jl]: I agree. I haven't been able to think of an elegant way of doing the same check.

Does anyone know the status/purpose of the "`\` revisions"?

File: base/linalg/conjarray.jl
Comment (andyferris) [inline base/linalg/conjarray.jl]: Right, I was hoping to address all the `StridedArray` behavior at some point... 

(it is a bit annoying that it is a type not a trait)

File: base/linalg/conjarray.jl
Comment (andyferris) [inline base/linalg/conjarray.jl]: Agreed

File: base/linalg/conjarray.jl
Comment (andyferris) [inline base/linalg/conjarray.jl]: In functionality, you are right, but this is a ambiguity resolution of the two methods above (not that `ConjArray{T} where T<:Real` is common, but it is possible for a user to create one, unless you prefer for the constructor to throw in that case?)

**Comment by andyferris** [comment]
> Should't there be some more specialized linear-algebra operations? e.g. x' * y should turn into a dot call.

I think this is done now. 

>  _conj(a) + _conj(b) should be the same as _conj(a + b) (similarly for *).

Hmm... I thought that getting rid of the `ConjArray` wrapper when creating a new array would be desirable. Are you suggesting it's more advantageous to take advantage of BLAS-level 1 routines? Shouldn't Julia generic code be pretty good at adding two vectors? (i.e. I assumed `a::AbstractArray + b::AbstractArray` would be calculated as `map(+, a, b)`)

**Comment by andyferris** [comment]
A question: what should `(v').'` and `(v.')'` become? Should it give a `ConjVector` or make a new conjugated copy of `v`?

The goal here was to make both `.'` and `'` return a view for vectors, but I hadn't defined `conj(v::AbstractVector) = ConjArray(v)`, so this seems a little at odds.

**Comment by timholy** [comment]
IIRC, the last time I looked Julia was as fast as BLAS for 1d operations (and indeed faster for small objects because of inlining, which is not possible with BLAS). There may have been some regression in our SIMD capabilities, so that may not be so true anymore, but I'd rather see this implemented in Julia code and then work on our SIMD.

File: test/linalg/conjarray.jl
Comment (tkelman) [inline test/linalg/conjarray.jl]: don't think this outer testset is necessary, the test system inserts a test set around each file anyway

Comment (andyferris) [inline test/linalg/conjarray.jl]: Right, thanks for the heads-up. The behavior of Base tests is quite different to what happens with packages, but I had seen an issue (from memory) about using `@testset` more so I used my usual style and hoped things would be picked up.

(is the hope to unify the package and base approaches in the future?)

Comment (tkelman) [inline test/linalg/conjarray.jl]: yes, https://github.com/JuliaLang/julia/issues/18740 and #19567

**Comment by andyferris** [comment]
Just saw [this](https://discourse.julialang.org/t/0-6-release-timeline/836/20?u=andyferris). Should we add a milestone for v0.6 for this PR? IMO, it is a (necessary?) refinement of #19670 (but it also adds a feature, so isn't just a bugfix).

**Comment by stevengj** [comment]
There are the BLAS3 cases of `conjmatrix * conjmatrix`, `conjtransposematrix * matrix`, etcetera, where will it certainly be faster to rewrite in terms of matrix multiplies on the parent matrices.

**Comment by stevengj** [comment]
For example, it would be good to have
```
B = A'
B * C, C * B
```
use BLAS if `A` and `C` are `Matrix{Float64}` or ``Matrix{Complex{Float64}}`.

File: test/linalg/conjarray.jl
Comment (andyferris) [inline test/linalg/conjarray.jl]: Very cool!

**Comment by andyferris** [comment]
Right, the next step is a much more disruptive (in terms of code) PR that introduces `TransposedMatrix`.

Because of the combinatorial explosion of methods of having all the `Ac_mul_B` methods *and* dispatch on all the transpose and conjugation types, at that point I'll reverse tack and start removing any and all `Ac_mul_B` methods I see and re-write it in terms of a submethod of `*(::ConjTransposedMatrix, ::AbstractMatrix)` (or vectors, etc). (At that point it is natural that we look at deprecating the special parsing of `a' * b`.)

I'm guessing this is post v0.6 (primarily for the reason is that we've already run out of time) unless people really think it's worth pursuing now.

File: test/sparse/sparse.jl
Comment (Sacha0) [inline test/sparse/sparse.jl]: Status of `\` revisions: Chances are no one has worked on improving the mechanism further since #16548 (particularly given that improving that mechanism depends on compiler improvements, see https://github.com/JuliaLang/julia/pull/16979#issuecomment-227935000).

Had a brief look at fixing this test. A simple fix would be to retrieve the method matches for `\` with `(SparseMatrixCSC, AbstractVecOrMat)` arguments and check that all such methods live in the `SparseArrays` module. For example, something along the lines of `all(meth -> meth.module == Base.SparseArrays, methods(\, (SparseMatrixCSC, AbstractVecOrMat)).ms)` could do the trick. Thoughts? Best!

Comment (andyferris) [inline test/sparse/sparse.jl]: Hmm, OK. At first I thought the issue might have been it going through a `rowvector.jl` method first before going to the `Sparse` module, but in fact this may be a disambiguity method that I add to `Sparse`, so that should work.

Comment (Sacha0) [inline test/sparse/sparse.jl]: > a disambiguity method that I added to Sparse

Correct. That there are now two matching methods rather than one is what caused the existing test to fail:
```julia
julia> which(\, (SparseMatrixCSC, AbstractVecOrMat))
ERROR: method match is ambiguous for the specified argument types
Stacktrace:
 [1] which(::Any, ::Any) at ./reflection.jl:715

julia> methods(\, (SparseMatrixCSC, AbstractVecOrMat))
# 2 methods for generic function "\":
\(::SparseMatrixCSC, ::RowVector) in Base.SparseArrays at sparse/linalg.jl:875
\(A::SparseMatrixCSC, B::Union{AbstractArray{T,1},AbstractArray{T,2}} where T) in Base.SparseArrays at sparse/linalg.jl:855
```
Best!

**Comment by andyferris** [comment]
OK I've done a little bit more work on this and rebased... just the sparse module `\` test to go, I think. That could be done separately or together, don't mind.

File: base/linalg/rowvector.jl
Comment (tkelman) [inline base/linalg/rowvector.jl]: loses

**Comment by StefanKarpinski** [comment]
There are three reviewers, none of which have officially approved this. @stevengj, @tkelman, @Sacha0 – can you approve if this is good to go, modulo typos?

File: test/sparse/sparse.jl
Comment (andyferris) [inline test/sparse/sparse.jl]: I fixed this for now, but I'm guessing that this will need to change again with `TransposedMatrix` (the thing it is testing may not even be true after that...).

**Comment by andyferris** [comment]
OK, I fixed the broken test (upgraded from "brittle" to "somewhat brittle") and typo, and have rebased.

Regarding the `StridedArray` interface functions (`pointer` and `stride`) - do people thin that these should be implemented here even though they won't be used, or should we fix/attack the `StridedArray` definition first (either by making it a trait or expanding its union)?

To me, this had felt like work that could be done along with `TransposedMatrix` - they will both be touching a similar set of function definitions (remember that `RowVector` simply unwraps itself for `*`, `\`, `/`, ..., but `TransposedMatrix` won't have that luxury if we get rid of `Ac_mul_B`, etc).

File: base/linalg/rowvector.jl
Comment (stevengj) [inline base/linalg/rowvector.jl]: Why does this make a copy?

File: base/linalg/rowvector.jl
Comment (stevengj) [inline base/linalg/rowvector.jl]: I'm worried about potential performance regressions if we don't also have `*(rowvec,Matrix) = At_mul_B` and `*(conjrowvec,Matrix) = Ac_mul_B`

**Comment by StefanKarpinski** [comment]
What do we need to do to make forward progress here?

File: base/linalg/rowvector.jl
Comment (andyferris) [inline base/linalg/rowvector.jl]: The first is true, but not the way you think: you end up on [this line](https://github.com/andyferris/julia/blob/9e0e1cf6641aa0c609c4efa3f4e8f9ccb0dbd536/base/linalg/rowvector.jl#L168) so you get `*(rowvec, matrix) = RowVector(At_mul_B(matrix, parent(rowvec)))`. Would that be a performance penalty?

I should chase down all the conjugated cases and make sure they do something sensible. I think what happens is a conjugated copy of the parent vector is created as a temporary, and the above gets called. (We should be able to remove this temporary).

Comment (stevengj) [inline base/linalg/rowvector.jl]: `RowVector(At_mul_B(matrix, parent(rowvec)))` should be fine.

File: base/linalg/rowvector.jl
Comment (andyferris) [inline base/linalg/rowvector.jl]: The logic was to make sure that vectors are always unwrapped, so that calls to BLAS are generally used. This is partly because the PR doesn't create a bunch of new methods for `*(conjarray, array)`, etc. We also have "view semantic" for row vectors and "copy semantic" for vectors, which seemed more consistent than views for vectors and rowvectors and copies for everything else.

(That is, from the user's perspective, only the new `RowVector` type gets the view semantic, conjugated or not. The old types behave as before, so *hopefully* this is more backward compatible. Ideally these changes would have been made at the same time as `TransposedMatrix` to make all of this more consistent...)

**Comment by andyferris** [comment]
* There might be a regression in `vec' * mat` such that an extra conjugated vector copy is made. Can I have feedback on whether people feel this will be a large or small problem, or could be addressed separately vs should be addressed in this PR? (This regression was introduced by #19670)

* Can I have a final word about providing the strided array functions (`pointer` and/or `stride`) for these views (given they wouldn't be used by code in `Base` at the moment)?

**Comment by stevengj** [comment]
We can't have `pointer` etc. for `ConjArray`, and I think we should punt on that for `RowVector` (given that `StridedArray` is not yet a trait anyway).

File: test/linalg/conjarray.jl
Comment (tkelman) [inline test/linalg/conjarray.jl]: 4-space

**Comment by andyferris** [comment]
Rebased, indentation fixed.

(If you want to merge this now before "alpha" then we can look at any remaining `Ac_mul_B` performance regressions where a `RowVector` is returned which were introduced in #19670 in a separate, more focused PR. I don't see such performance penalties as being very large or very difficult to fix.)

**Comment by andyferris** [comment]
Bump

**Comment by tkelman** [comment]
I like it. If `ConjMatrix` isn't going to occur in normal operations yet, should we maybe hold off on exporting that alias for now?

**Comment by andyferris** [comment]
> I like it. 

Thanks. I'm already toying with `TransposedMatrix` and refactoring matmul, etc. I think we'll see nice simplifications out of that.

> If `ConjMatrix` isn't going to occur in normal operations yet, should we maybe hold off on exporting that alias for now?

Right. On one hand, I'm definitely already using it in my toying around, but it is one more symbol in the namespace. I kind-of felt having the vector and array forms and not the matrix form as being a bit odd and distinct to how these things typically come in triples. I'm easy.

**Comment by stevengj** [comment]
LGTM.

I'd still like a method to improve performance in `y = x'; x * A`, but that can be a separate PR.

**Review by stevengj — APPROVED** [review]

File: base/linalg/conjarray.jl
Comment (KristofferC) [inline base/linalg/conjarray.jl]: Perhaps a doctest here with an example of behaviour?

File: base/linalg/rowvector.jl
Comment (KristofferC) [inline base/linalg/rowvector.jl]: Perhaps a `@ref` to `ConjArray` and maybe a doctest?

File: base/linalg/conjarray.jl
Comment (andyferris) [inline base/linalg/conjarray.jl]: I'll have to learn what a doctest is, first. I guess I'll try the docs.

File: base/linalg/rowvector.jl
Comment (andyferris) [inline base/linalg/rowvector.jl]: I didn't know about `@ref` either, but that seems useful.

**Comment by andyferris** [comment]
Done, @KristofferC . I had trouble getting either of
```
[`'`](@ref)
[`.'](@ref)
```
to work, though, but perhaps this is good enough.

**Comment by andyferris** [comment]
I think this might be good to go?

**Comment by stevengj** [comment]
Looks like the tests need to be updated after #20423.

**Comment by StefanKarpinski** [comment]
Bump. This is one of the last remaining feature changes for 0.6.

**Comment by andyferris** [comment]
@StefanKarpinski waiting on CI, but this should be good to merge now.

File: test/linalg/rowvector.jl
Comment (tkelman) [inline test/linalg/rowvector.jl]: shouldn't delete these

Comment (andyferris) [inline test/linalg/rowvector.jl]: thanks

Comment (andyferris) [inline test/linalg/rowvector.jl]: I think one of the rebases must have done something odd here. I'll fix it.

Comment (andyferris) [inline test/linalg/rowvector.jl]: fixed

**Comment by andyferris** [comment]
Squashed and rebased, for convenience.

**Comment by StefanKarpinski** [comment]
So the pending work here is all adding optimized internal methods, right?

**Comment by andyferris** [comment]
That's correct, Stefan. It shouldn't affect the user interface or behaviour.

**Comment by StefanKarpinski** [comment]
Great – let's pull the trigger once CI passes everywhere.

**Comment by andyferris** [comment]
Looks like Travis finally started.

BTW, Stefan, I have started on a branch [here](https://github.com/andyferris/julia/tree/ajf/transposedmatrix) with `TransposedMatrix` and a refactoring (i.e. removal) of all the `Ac_mul_B` and `Ac_mul_B!` methods into `*` and `A_mul_B!`. So far I've attacked the generic and strided/BLAS multiplications, but that leaves symmetric, hermitian, diagonal, bidiagonal, triangular, sparse, etc, as well as the division methods... :) I've also made some movements of code so that BLAS is more modular. (I've noticed there are assumptions everywhere that the external libraries are present and it will take quite a bit of effort to turn BLAS, LAPACK, ARPACK, etc into proper modules that can be loaded at the end of `LinAlg` - ideally, optionally, with the caveat that there aren't native implementations of many things.)

File: base/linalg/rowvector.jl
Comment (tkelman) [inline base/linalg/rowvector.jl]: not worth going through CI again right now, but this should probably say `conj(v::RowVector)` to be specific?

Comment (pabloferz) [inline base/linalg/rowvector.jl]: I took the liberty of changing this over #20446

Comment (andyferris) [inline base/linalg/rowvector.jl]: @tkelman are we moving towards these more accurate type signatures in docs? I was copying the style that I've seen in a lot of other places, but for myself I would have more naturally written it like you have here.

Comment (tkelman) [inline base/linalg/rowvector.jl]: if the docstring description applies generally, then the signature should be general. if it's about a method for a specific type, then the signature should show that

Comment (andyferris) [inline base/linalg/rowvector.jl]: thanks

**Pull request merged**
